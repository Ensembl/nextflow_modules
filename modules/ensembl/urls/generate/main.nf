process URLS_GENERATE {
    tag "$meta.id"
    label 'process_low'

    input:
    tuple val(meta), val(protocol), val(user), val(host), val(port), val(password), val(database)

    output:
    tuple val(meta), env(base_url) , emit: url_sql
    path "versions.yml"           , emit: versions

    script:

    """
    protocol="${protocol}"
    user="${user}"
    host="${host}"
    port="${port}"
    password="${password}"
    database="${database}"

    # Start constructing the URL
    base_url="\${protocol}://\${user}@\${host}:\${port}"

    # Add password if it is not an empty string
    if [ "\$password" != "" ]; then
        base_url="\${protocol}://\${user}:\${password}@\${host}:\${port}"
    fi

    # Add database if it is not an empty string
    if [ "\$database" != "" ]; then
        base_url="\${base_url}/\${database}"
    fi

    export  base_url

    echo -e -n "${task.process}:\n\tensembl-genomio: " > versions.yml
    """
    
    stub:
    """
    host = host.com
    user = user
    port = 3306

    echo -e -n "${task.process}:\n\tensembl-genomio: " > versions.yml
    """
}
