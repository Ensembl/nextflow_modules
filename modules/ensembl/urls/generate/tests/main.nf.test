nextflow_process {

    name "Test Process URLS_GENERATE"
    script "../main.nf"
    process "URLS_GENERATE"

    tag "modules"
    tag "modules_ensembl"
    tag "urls"
    tag "urls/generate"
    
    test("Should generate MySQL URL correctly with all parameters") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test1' ],
                    'mysql',
                    'user1',
                    'host1.com',
                    '3306',
                    'password1',
                    'database1'
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.url_sql != null },
                { assert process.out.versions != null },
                { assert process.out.url_sql.get(0).get(1) == 'mysql://user1:password1@host1.com:3306/database1' },
                { assert snapshot(process.out.url_sql).match("generated_url_full") },
                { assert snapshot(process.out.versions).match("versions_full") }
            )
        }
    }

    test("Should generate URL without password") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test2' ],
                    'postgresql',
                    'user2',
                    'host2.com',
                    '5432',
                    '',
                    'database2'
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.url_sql != null },
                { assert process.out.versions != null },
                { assert process.out.url_sql.get(0).get(1) == 'postgresql://user2@host2.com:5432/database2' },
                { assert snapshot(process.out.url_sql).match("generated_url_no_password") },
                { assert snapshot(process.out.versions).match("versions_no_password") }
            )
        }
    }

    test("Should generate URL without database") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test3' ],
                    'oracle',
                    'user3',
                    'host3.com',
                    '1521',
                    'password3',
                    ''
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.url_sql != null },
                { assert process.out.versions != null },
                { assert process.out.url_sql.get(0).get(1) == 'oracle://user3:password3@host3.com:1521' },
                { assert snapshot(process.out.url_sql).match("generated_url_no_database") },
                { assert snapshot(process.out.versions).match("versions_no_database") }
            )
        }
    }
}