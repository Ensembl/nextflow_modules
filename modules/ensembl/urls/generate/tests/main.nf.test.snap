{
    "versions_no_database": {
        "content": [
            [
                "versions.yml:md5,d817b1f3c9f2bd1f04c668984320951c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-06-17T11:31:22.822012541"
    },
    "versions_full": {
        "content": [
            [
                "versions.yml:md5,d817b1f3c9f2bd1f04c668984320951c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-06-17T11:30:59.208840884"
    },
    "versions_no_password": {
        "content": [
            [
                "versions.yml:md5,d817b1f3c9f2bd1f04c668984320951c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-06-17T11:31:11.312214413"
    },
    "generated_url_full": {
        "content": [
            [
                [
                    {
                        "id": "test1"
                    },
                    "mysql://user1:password1@host1.com:3306/database1"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T15:48:40.650816"
    },
    "generated_url_no_password": {
        "content": [
            [
                [
                    {
                        "id": "test2"
                    },
                    "postgresql://user2@host2.com:5432/database2"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T15:48:42.651032"
    },
    "generated_url_no_database": {
        "content": [
            [
                [
                    {
                        "id": "test3"
                    },
                    "oracle://user3:password3@host3.com:1521"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T15:48:44.600396"
    }
}